### PR TITLE
[Final Project] ChocoPy2Bril: Translating ChocoPy to Bril

### LINKED ISSUE #603: Project Proposal: ChocoPy2Bril
**What will you do?**
The primary goal is to create a compiler frontend, named "ChocoPy2Bril," that translates a core subset of the ChocoPy language directly into the Bril representation. This project will focus on the AST traversal and code generation phase by leveraging an existing ChocoPy parser to handle lexical analysis and syntax analysis.

Target ChocoPy Subset:

- Primitive Types: Integers and Booleans (int, bool).
- Basic Operations: Arithmetic (+, -, *, /, %), Relational (==, !=, <, >, <=, >=), and Logical (and, or, not).
- Variables: Local variable declaration and assignment.
- Control Flow: Conditional statements (if/else) and simple iteration (while loops).
- Functions: Function definition, argument passing, and return statements.

**How will you do it?**
We will use an existing parser/lexer that builds an AST representation of ChocoPy, and we will go in and implement the translation from AST to Bril. For the existing parser/lexer, we are looking at either finding an existing Python parser/lexer, or we will use an existing [ChocoPy parser/lexer](https://github.com/gabe-raulet/chocopy2bril) to build the AST. Which route we take will ultimately come down to whichever linking is easier.

**How will you empirically measure success?**
We will empirically measure success by manually creating a test suite of about 20 ChocoPy programs (targeting the above functionality subset and edge cases) and compare the program output of our ChocoPy2Bril implementation with the existing ChocoPy implementation, and ensure that there are no differences. 

**Team members:**
@Jacqueline-Wen @SerenaYZhang @magg1egao

### PR DESCRIPTION
**Pull request by Jacqueline-Wen**
Closes #603

### CODE REVIEW

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+"""
+[[extra.authors]]
+name = "Jacqueline Wen"
+[[extra.authors]]
+name = "Maggie Gao"
+[[extra.authors]]
+name = "Serena Zhang"
++++
+
+### What was the goal?
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: Because this post is supposedly for external consumption, there's no need for headings that correspond 1:1 to my suggestions for what to include. For example, the text here works just fine as an introduction and probably doesn't need a heading at all.

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+The primary goal is to create a compiler frontend, named "ChocoPy2Bril," that translates a core subset of the ChocoPy language directly into the Bril representation. This project will focus on the AST traversal and code generation phase by leveraging an existing ChocoPy parser to handle lexical analysis and syntax analysis.
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: Can you please include a link to the ChocoPy homepage?

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
...ficulty, we were able to familiarize ourselves with ChocoPy's grammar, and we had to settle on how we were planning on implementing the translation from ChocoPy to Bril. At the recommendation of Professor Sampson, we focused on efforts on the AST translation, instead of actually building the AST from scratch, as this would be a really tedious, time consuming, and fairly non-trivial task to do. This allowed us to focus on the code generation phase of the compiler, which was our main goal. The next step in the design would be to decide how we were going to generating the AST from the ChocoPy program. Originally, we thought that the simplest solution would be to use the built in Python AST module to generate the AST from the ChocoPy program. However, we soon realized that there were small nuances between the ChocoPy grammar and the Python grammar, which made this approach infeasible. After some research, we found that there were multiple existing ChocoPy parser implemented in Python, which we were able to leverage to generate the AST for us. We ended up deciding to use "yangdanny97/chocopy-python-compiler" as our ChocoPy parser, as it was fairly well documented and seemed easy to use.
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: When you talk about the existing parser, I think it would be great to:

* include a link
* make up a name, like "Danny Yang's parser," so you can refer to this more easily instead of with the full GitHub path

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
...ormat. As a result, we just had to hard code the translation between the ChocoPy AST nodes and the Bril instructions. Most notably, we created a `CodeState` class to keep track of temporary variables and variable types. This helps generate unique temporary variables names during the translation. The `translate_prog` function serves to handle function definitions, the ChocoPy main function body, as well as any global variable declarations. `translate_func_def` is responsible for mapping arguments and local variable definitions, translating statements, and adding a default return value if none is proveded. `translate_var_def` translates variable definitions, and `translate_expr` handles the translation of instructions, including literals, identifiers, binary operations, unary operations, and function calls. Finally, `translate_stmt` translates statements such as assignment, if statements, while loops, return statements, and expression statements including print. The main function orchestrates the overall AST translation, including calling "yangdanny97/chocopy-python-compiler" to create the ChocoPy AST, as well as calling the appropriate functions from above to generate the Bril code.
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: This is a very low-level description of the individual functions in your implementation. That's fine if that's what you have to say, but if you want, you can consider dropping the references to particular functions (the reader probably doesn't care what you named your functions) and attempting to give a higher-level insight into what your overall strategy was.

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
...a test suite of 20 unit tests that covered the various features we implemented in our ChocoPy2Bril translator. Each test case consisted of a ChocoPy program, and we wrote a script that verified that the result of running the generated Bril code matched the ChocoPy program's output. This helped us ensure the correctness of our implementation and catch any bugs or edge cases. Our test suites were partially adapted/inspired from the test suites provided in the "yangdanny97/chocopy-python-compiler" repository. Initially, we wanted to directly use tests from "yangdanny97/chocopy-python-compiler". However, we soon realized that most of those tests were fairly complex and covered a lot of ChocoPy features that we were not planning on implementing. As a result, we had to edit them to fit our implemented features, as well as manually write some of our tests from scratch. Additionally, on the GitHub issue for this project, we were intially planning on using Turnt for testing. However, we soon realized that since the way that we were checking correctness/calling functions differed across test cases, it was simply easier to write our own testing script that handled each test case individually.
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: > on the GitHub issue for this project

Please include a link.

Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: > Each test case consisted of a ChocoPy program, and we wrote a script that verified that the result of running the generated Bril code matched the ChocoPy program's output.

Where did your "ground truth" for the ChocoPy program's output come from?

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+is_even_x: bool = False
+is_even_y: bool = False
+is_even_x = is_even(x)
+is_even_y = is_even(y)
+print(is_even_x)
+print(is_even_y)
+```
+
+When we run our ChocoPy2Bril translator on this test case, it generates the following Bril code:
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: I don't think it's very helpful to include the full Bril listing inline, unless there's something specific you want readers to look at in this listing (in which case you can say that).

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 5 above)
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: > However, we soon realized that since the way that we were checking correctness/calling functions differed across test cases

I don't quite understand this: isn't the correctness check always "the compiled Bril program prints the same thing as the original ChocoPy program"? In that case, what are the relevant differences? Could you maybe say a little bit more about the nuances of setting these things up to test them?

File: content/blog/2025-12-16-ChocoPy2Bril.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+  _temp8: bool = id is_even_x;
+  print _temp8;
+  _temp9: bool = id is_even_y;
+  print _temp9;
+  ret;
+}
+```
+
+### What were the hardest parts to get right?
+The hardest part of the project was definitely the debugging. It was often difficult to trace back errors in the generated Bril code to the corresponding ChocoPy AST nodes. This made it challenging to identify and fix bugs in our implementation. When the test results did not match, we had to analyze whether it was our code that was incorrect, or if there was an issue with the ChocoPy parser we were using. Additionally, there were moments where our Bril file translation would fail, and we had to inspect whether it was due to our translation logic or if there were limitations in the Bril language itself.
```
Comment (sampsyo) [inline content/blog/2025-12-16-ChocoPy2Bril.md]: Could I ask you to include at least one example of such a bug that you found using your test suite? It would go a long way toward illustrating this notion you're describing here re. the root causes (e.g., I don't fully see what bugs "limitations in the Bril language itself" would cause).

**Review by sampsyo — CHANGES_REQUESTED** [review]
Hi, everybody—congratulations on finishing up. It sounds like this ended up working out adequately, even if the language subset was somewhat small.

I have a few writing suggestions—could you please address those and then use GitHub's "request review" button to let me know when you're ready to publish?

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 1 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: I changed it!

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 2 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: Done!

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 3 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: Done!

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 4 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: I think we don't mind including more details!

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 5 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: Done!

Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: Sorry for the ambiguous language. We wrote a script that ran both the generated Bril code and the ChocoPy program, and verified that the outputs were equivalent. I changed the wording.

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 6 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: I removed it.

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 5 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: I included this in the newest commit: 
One notable difference was that Bril outputted true and false as lowercase, while ChocoPy outputted them as capitalized. As a result, a simple `diff` is not enough to verify equality. Additionally, while some of our test cases had a main function, others did not. Since we ran our ChocoPy programs on the command line via Python, we had to manually call the main function in the testing script (the syntax of ChocoPy's main function differs from Python).

File: content/blog/2025-12-16-ChocoPy2Bril.md (hunk 7 above)
Comment (Jacqueline-Wen) [reply content/blog/2025-12-16-ChocoPy2Bril.md]: I included this in the newest commit:
(ex. trying to use modulo in Bril, when the language itself does not support that arithmetic operation)

**Comment by sampsyo** [comment]
Thanks for the additional details! This looks great.
